### PR TITLE
[SPARK-13049] Add First/last with ignore nulls to functions.scala

### PR DESCRIPTION
**Pull request by hvanhovell**
This PR adds the ability to specify the `ignoreNulls` option to the functions dsl, e.g: 
`df.select($"id", last($"value", ignoreNulls = true).over(Window.partitionBy($"id").orderBy($"other"))`

This PR is some where between a bug fix (see the JIRA) and a new feature. I am not sure if we should backport to 1.6.

cc @yhuai

### CODE REVIEW

**Comment by yhuai** [comment]
@hvanhovell Thanks for the PR. Do you know why expr/callUDF does not work?

**Comment by hvanhovell** [comment]
@yhuai ```expr("last(r, true)")``` would return an ```UnresolvedFunction(UnresolvedAttribute(r), Literal(true))```. The problem is that the ```WindowSpec``` does not recognize ```UnresolvedFunction```'s.

This is the cleaner fix. We could also add a match to the ```WindowSpec``` function for unresolved functions.

**Comment by hvanhovell** [comment]
retest this please

**Comment by hvanhovell** [comment]
retest this please

**Comment by rxin** [comment]
Why might this be a bug fix?

**Comment by hvanhovell** [comment]
A user is trying to get this working on 1.6 using the dataframe api. That doesn't work directly because functions.scala misses  the functions implemented in this PR. The indirect approach using ```expr(...)``` doesn't work because ```WindowSpec``` does not support ```UnresolvedFunctions```. 

I guess this is more a feature than a bug fix....

**Comment by rxin** [comment]
Actually can you update the Python API as well?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -349,19 +349,41 @@ object functions extends LegacyFunctions {
...
 
   /**
-   * Aggregate function: returns the first value in a group.
-   *
-   * @group agg_funcs
-   * @since 1.3.0
-   */
-  def first(e: Column): Column = withAggregateFunction { new First(e.expr) }
+    * Aggregate function: returns the first value in a group. The function does not consider null
+    * values when the ignoreNulls flag is set to true.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Can you write something like this to be more clear? And update all the docs (including Python).

"The function by default includes the first value it sees. When ignoreNulls is set to true, then it ignores the null values and includes the first non-null value. If all values are null, then null is returned."

**Comment by rxin** [comment]
Thanks - only some minor comment on the documentation to make it more clear.

**Comment by rxin** [comment]
Thanks - merging this in master.

**Pull request closed without merging**